presto-kinesis/src/main/java/io/prestosql/plugin/kinesis/KinesisShardCheckpointer.java

```diff
@@ -0,0 +1,176 @@
+import com.google.common.base.Throwables;
+import io.airlift.log.Logger;
+
+import static com.google.common.base.Throwables.throwIfUnchecked;
+
+public class KinesisShardCheckpointer
+{
+    private static final Logger log = Logger.get(KinesisShardCheckpointer.class);
+    private KinesisClientLeaseManager leaseManager;
+    private KinesisSplit kinesisSplit;
+    private String logicalProcessName;
+    private int currentIterationNumber;
+    private long dynamoReadCapacity;
+    private long dynamoWriteCapacity;
+    private KinesisClientLease kinesisClientLease;
+    private long checkpointIntervalMs;
+    private long nextCheckpointTimeMs;
+
+    public KinesisShardCheckpointer(
+            AmazonDynamoDB dynamoDBClient,
+            String dynamoDBTable,
+            KinesisSplit kinesisSplit,
+            String logicalProcessName,
+            int currentIterationNumber,
+            long checkpointIntervalMS,
+            long dynamoReadCapacity,
+            long dynamoWriteCapacity)
+    {
+        this(new KinesisClientLeaseManager(dynamoDBTable, dynamoDBClient),
+                kinesisSplit,
+                logicalProcessName,
+                currentIterationNumber,
+                checkpointIntervalMS,
+                dynamoReadCapacity,
+                dynamoWriteCapacity);
+    }
+
+    public KinesisShardCheckpointer(
+            KinesisClientLeaseManager leaseManager,
+            KinesisSplit kinesisSplit,
+            String logicalProcessName,
+            int currentIterationNumber,
+            long checkpointIntervalMS,
+            long dynamoReadCapacity,
+            long dynamoWriteCapacity)
+    {
+        this.leaseManager = leaseManager;
+        this.kinesisSplit = kinesisSplit;
+        this.logicalProcessName = logicalProcessName;
+        this.currentIterationNumber = currentIterationNumber;
+        this.checkpointIntervalMs = checkpointIntervalMS;
+        this.dynamoReadCapacity = dynamoReadCapacity;
+        this.dynamoWriteCapacity = dynamoWriteCapacity;
+
+        try {
+            this.leaseManager.createLeaseTableIfNotExists(this.dynamoReadCapacity, this.dynamoWriteCapacity);
+
+            KinesisClientLease oldLease = this.leaseManager.getLease(createCheckpointKey(currentIterationNumber));
+            if (oldLease != null) {
+                this.kinesisClientLease = oldLease;
+            }
+            else {
+                this.kinesisClientLease = new KinesisClientLease();
+                this.kinesisClientLease.setLeaseKey(createCheckpointKey(currentIterationNumber));
+            }
+        }
+        catch (ProvisionedThroughputException | InvalidStateException | DependencyException e) {
+            throwIfUnchecked(e);
+            throw new RuntimeException(e);
+        }
+        resetNextCheckpointTime();
+    }
+
+    private void resetNextCheckpointTime()
+    {
+        nextCheckpointTimeMs = System.currentTimeMillis() + checkpointIntervalMs;
```

Comment:
Changed this to nanotime, at a couple of other places where it is still System.currentTimeMillis is because we want to pass "Date" from which to fetch records, so nanoTime won't work there as per my understanding